src/common/tracer.cc

```diff
@@ -1,86 +1,62 @@
-#define dout_subsys ceph_subsys_osd
-#undef dout_prefix
-#define dout_prefix *_dout << "jaeger_tracing "
-
-namespace jaeger_tracing {
-
-  std::shared_ptr<opentracing::v3::Tracer> tracer = nullptr;
-
-  void init_tracer(const char* tracer_name) {
-    if (!tracer) {
-	YAML::Node yaml;
-	try{
-	  yaml = YAML::LoadFile("../src/jaegertracing/config.yml");
-	dout(3) << "yaml loaded" << yaml << dendl;
-	}
-	catch(std::exception &e) {
-	  dout(3) << "failed to load yaml file using default config" << dendl;
-	  auto yaml_config = R"cfg(
-disabled: false
-reporter:
-    logSpans: false
-    queueSize: 100
-    bufferFlushInterval: 10
-sampler:
-    type: const
-    param: 1
-headers:
-    jaegerDebugHeader: debug-id
-    jaegerBaggageHeader: baggage
-    TraceContextHeaderName: trace-id
-baggage_restrictions:
-    denyBaggageOnInitializationFailure: false
-    refreshInterval: 60
-)cfg";
-	yaml = YAML::Load(yaml_config);
-	dout(3) << "yaml loaded" << yaml << dendl;
-      }
-      static auto configuration = jaegertracing::Config::parse(yaml);
-      tracer = jaegertracing::Tracer::make( tracer_name, configuration,
-	    jaegertracing::logging::consoleLogger());
-      dout(3) << "tracer_jaeger init successful" << dendl;
-    }
-    //incase of stale tracer, configure with a new global tracer
-    if (opentracing::Tracer::Global() != tracer) {
-      opentracing::Tracer::InitGlobal(
-	  std::static_pointer_cast<opentracing::Tracer>(tracer));
-    }
-  }
+#include "common/ceph_context.h"
+#include "global/global_context.h"
+#include "common/tracer.h"
 
-  jspan new_span(const char* span_name) {
-    return opentracing::Tracer::Global()->StartSpan(span_name);
- }
+#ifdef HAVE_JAEGER
 
-  jspan child_span(const char* span_name, const jspan& parent_span) {
-    //no parent check if parent not found span will still be constructed
-    return opentracing::Tracer::Global()->StartSpan(span_name,
-	{opentracing::ChildOf(&parent_span->context())});
- }
+namespace jaeger_configuration {
 
-  void finish_span(const jspan& span) {
-    if (span) {
-      span->Finish();
-    }
+jaegertracing::samplers::Config const_sampler("const", 1, "", 0, jaegertracing::samplers::Config::defaultSamplingRefreshInterval());
+
+jaegertracing::reporters::Config reporter_default_config(jaegertracing::reporters::Config::kDefaultQueueSize, jaegertracing::reporters::Config::defaultBufferFlushInterval(), true, jaegertracing::reporters::Config::kDefaultLocalAgentHostPort, "");
+
+jaegertracing::propagation::HeadersConfig headers_config("", "", "", "");
+
+jaegertracing::baggage::RestrictionsConfig baggage_config(false, "", std::chrono::steady_clock::duration());
+
+jaegertracing::Config jaeger_rgw_config(false, const_sampler, reporter_default_config, headers_config, baggage_config, "rgw", std::vector<jaegertracing::Tag>());
+
+jaegertracing::Config jaeger_osd_config(false, const_sampler, reporter_default_config, headers_config, baggage_config, "osd", std::vector<jaegertracing::Tag>());
```

Comment:
these shouldn't be in a common file. the rgw config should be in `rgw/rgw_tracer.*`, and the osd should have its own too